docs-2.0/1.introduction/3.nebula-graph-architecture/4.storage-service.md

```diff
@@ -1,11 +1,261 @@
+
+Failure: If the system has only one replica, Leader will be itself. If failure happens, the system will be completely unavailable. If the system has three replicas, one of them will be Leader and the rest will be Followers. If the Leader fails, the rest two can still vote for a new Leader (and a Follower), and the system is still available. But if any of the two Followers fails again, the system will be completely unavailable due to inadequate voters.
+
+!!! Note
+
+    Raft and HDFS have different modes of duplication. Raft is based on a majority vote, so the number of replicas cannot be even.
+
+Listener: As is a special role in Raft, it cannot vote or keep data consistency. In Nebula Graph, it reads Raft-wal from Leader and synchronizes it to ElasticSearch cluster.
+
+### Multi Group Raft
+
+Storage service supports a distributed cluster architecture, so Nebula Graph implements Multi Group Raft according to Raft requirement. Each Raft group stores all the replicas of each partition. One replica is the leader, while others are followers. In this way, Nebula Graph achieves strong consistency and high availability. The functions of Raft are as follows.
+
+Nebula Graph uses Multi Group Raft to improve performance when there are many partitions because Raft-wal cannot be NULL. When there are too many partitions, costs will increase, such as storage information in Raft group, WAL files, or batch operation in low load.
+
+There are two key points to implement the Multi Raft Group:
+
+- To share transport layer
+
+   Each Raft Group sends messages to its corresponding peers. So if the transport layer cannot be shared, the connection costs will be very high.
+
+- To share thread pool
+
+  Raft Groups share the same thread pool to prevent starting too many threads and a high context switch cost.
+
+### Batch
+
+For each partition, it is necessary to do a batch to improve throughput when writing the WAL serially. As Nebula Graph uses WAL to implement some special functions, batches need to be grouped, which is a feature of Nebula Graph.
+
+For example, lock-free CAS operations will execute after all the previous WALs are committed. So for a batch, if there are several WALs in CAS type, we need to divide this batch into several smaller groups and make sure they are committed serially.
+
+### Listener
+
+Listener is designed for **storage expansion**. It takes a long time for the newly added machines to synchronize data during expansion. If synchronizing as a follower, the availability of the entire cluster will decrease.
+
+Listener will be written into the WAL command. If the leader finds a command of `add learner` when writing the WAL, it will add the learner to its peers and mark it as a Listener. Listeners will not be included into the majorities, but logs will still be sent to them as usual. Listeners themselves will not initiate elections.
```

Comment:
The listener will write into the command WAL.

Reply:
t will add the learner to its peers
it will add the  listener  to its peers

Reply:
Listeners will not be included into the majorities. -> Listeners can not join the quorum votes.

Reply:
I think "it will add the learner to its peers" is right, considering the context.